contrib/replicate-build/template/predict.py

```diff
@@ -0,0 +1,135 @@
+                orig_ids[next_local_id] = args.get('stream_id')
+                open_ids.add(next_local_id)
+                args['stream_id'] = next_local_id
+                args['model'] = 'checkpoints' # ignore specified model and substitute our own
+                next_local_id += 1
+                await session.handle(cmd, args) # enqueue the command
+
+            # now, we need to retrieve results...
+            while True:
+                (response_type, response_data) = await token_queue.get()
+                local_id = response_data.get('stream_id')
+                orig_id = orig_ids.get(local_id)
+                if not local_id in open_ids:
+                    if not local_id in prev_warned:
+                        print(f'WARNING: Ignoring extra content for {local_id!r} (formerly {orig_id!r})', file=sys.stderr)
+                        prev_warned.add(local_id)
+                    continue
+
+                if response_type != 'TOKEN':
+                    print(f'WARNING: Unrecognized response type {response_type!r}; may need to be updated', file=sys.stderr)
+
+                response_data['stream_id'] = orig_id
+
+                yield f'{response_type!s} {json.dumps(response_data)}\n'
+                if response_data.get("finish_reason") is not None:
+                    open_ids.discard(local_id)
+                if "error" in response_data:
+                    if local_id is not None:
+                        # Treat this as an error only for the one stream, not the whole result
+                        open_ids.discard(local_id)
+                    else:
+                        # an error without an id attached means we're finished
+                        return
+                if len(open_ids) == 0:
+                    return
+        finally:
+            session.close()
+
+    # Note that cog invokes this with apply_async in a multiprocessing threadpool.
+    def predict(self, ops_batch_json: str) -> cog.ConcatenateIterator[str]:
+        # if input doesn't parse, we want to fail fast; so do this first
+        ops_batch = json.loads(ops_batch_json)
+
+        loop = asyncio.new_event_loop()
+        loop.set_debug(True)
+        asyncio.set_event_loop(loop)
+
+        gen = self.lmtp_inference_stream(ops_batch)
+        try:
+            while True:
+                yield loop.run_until_complete(gen.__anext__())
+        except StopAsyncIteration:
+            return
+        finally:
+            loop.run_until_complete(loop.shutdown_asyncgens())
+            loop.close()
+
+if __name__ == '__main__':
+    if len(sys.argv) == 2:
+        request = sys.argv[1]
+    else:
+        request = sys.stdin.read()
+    args = json.loads(request)
+    p = Predictor()
+    print('INFO: About to run setup()...', file=sys.stderr)
+    p.setup()
+    print('INFO: Finished with setup(), about to start predictions', file=sys.stderr)
+    for result in p.predict(request):
+        print(result, flush=True)
```

Comment:
Can this be run outside of replicate on a local dev machine? I would very much appreciate a quick way to test things, when we change things about the LMTP scheduler or other components.

Reply:
Absolutely, I ran it locally during testing a lot. There are two ways to do it; one is to have Cog build a Docker image, and is more true to the real-world use case:

```
cog predict -i ops_batch_json='[["GENERATE",{...}],["GENERATE",{...}]]'
```

...and the other is to run it directly:

```
python predict.py '[["GENERATE", {...}],["GENERATE", {...}]]'
```